pkg/apis/autoscaling/v2alpha1/types.go

```diff
@@ -0,0 +1,103 @@
+}
+
+// Specification of ClusterAutoscaler. Empty for for now.
+type ClusterAutoscalerSpec struct {
+}
+
+// Type of ClusterAutoscalerCondition
+type ClusterAutoscalerConditionType string
+
+const (
+	// Condition that explains what is the current health of ClusterAutoscaler or its node groups.
+	ClusterAutoscalerHealth ClusterAutoscalerConditionType = "Health"
+	// Condition that explains what is the current status of a node group with regard to
+	// scale down activities.
+	ClusterAutoscalerScaleDown ClusterAutoscalerConditionType = "ScaleDown"
+	// Condition that explains what is the current status of a node group with regard to
+	// scale down activities.
+	ClusterAutoscalerScaleUp ClusterAutoscalerConditionType = "ScaleUp"
+)
+
+// Status of ClusterAutoscalerCondition.
+type ClusterAutoscalerConditionStatus string
+
+const (
+	// Statuses for Health condition type.
+	ClusterAutoscalerHealthy   ClusterAutoscalerConditionStatus = "Healthy"
+	ClusterAutoscalerUnhealthy ClusterAutoscalerConditionStatus = "Unhealthy"
+
+	// Statuses for ScaleDown condition type.
+	ClusterAutoscalerCandidatesPresent ClusterAutoscalerConditionStatus = "CandidatesPresent"
+	ClusterAutoscalerNoCandidates      ClusterAutoscalerConditionStatus = "NoCandidates"
+
+	// Statuses for ScaleUp condition type.
+	ClusterAutoscalerCandidatesNeeded ClusterAutoscalerConditionStatus = "Needed"
+	ClusterAutoscalerNotNeeded        ClusterAutoscalerConditionStatus = "NotNeeded"
+	ClusterAutoscalerInProgress       ClusterAutoscalerConditionStatus = "InProgress"
+	ClusterAutoscalerNoActivity       ClusterAutoscalerConditionStatus = "NoActivity"
+)
+
+// ClusterAutoscalerCondition describes some aspect of ClusterAutoscaler work.
+type ClusterAutoscalerCondition struct {
+	// Defines the aspect that the condition describes. For example, it can be Health or ScaleUp/Down activity.
+	Type ClusterAutoscalerConditionType `json:"type,omitempty" protobuf:"bytes,1,opt,name=type"`
+	// Status of the condition. Tells how given aspect
+	Status ClusterAutoscalerConditionStatus `json:"status,omitempty" protobuf:"bytes,2,opt,name=status"`
+	// Free text extra information about the condition. It may contain some
+	// extra debugging data, like why the cluster is unhealthy.
+	Message string `json:"message,omitempty" protobuf:"bytes,3,opt,name=message"`
+	// Since when the condition was in the given state.
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty" protobuf:"bytes,4,opt,name=lastTransitionTime"`
+}
+
+// Status of ClusterAutoscaler
+type ClusterAutoscalerStatus struct {
+	// Status information of individual node groups on which CA works.
+	NodeGroupStatuses []NodeGroupsStatus
+	// Conditions that apply to the whole autoscaler.
+	ClusterwideConditions []ClusterAutoscalerCondition
+}
+
+// Status of a group of nodes controlled by ClusterAutoscaler.
+type NodeGroupStatus struct {
+	// Name of the node group. On GCE it will be equal to MIG url, on AWS it will be ASG name, etc.
```

Comment:
Well the problem here is that AWS autoscaling groups don't have URLs (and Azure's equivalent probably too). On GCE/GKE name may be not unique. I can call it and Id but then it may be even more confusing.